### PR TITLE
Support (fieldless) enums in `Config::Input`

### PR DESCRIPTION
**Pull request by Vrixyz**
Improves #40.

- in bytemuck, `Pod` trait was split in multiple traits in release 1.9.0 https://github.com/Lokathor/bytemuck/blob/main/changelog.md#190
- NoPadding renamed into NoUninit: https://github.com/Lokathor/bytemuck/pull/94
- There is no benchmarks, I think the impact should be very minimal, but I figured I would bring it up.

### CODE REVIEW

File: src/frame_info.rs [hunk 1]
```diff
@@ -23,19 +23,32 @@ impl<S: Clone> Default for GameState<S> {
...
 #[derive(Debug, Copy, Clone, PartialEq)]
 pub(crate) struct PlayerInput<I>
 where
-    I: Copy + Clone + PartialEq + bytemuck::Pod + bytemuck::Zeroable,
+    I: Copy
+        + Clone
+        + PartialEq
+        + bytemuck::NoUninit
+        + bytemuck::CheckedBitPattern
+        + bytemuck::Zeroable,
```
Comment (Vrixyz) [inline src/frame_info.rs:35]: Not sure PlayerInput needs all those type constraints, as tests are passing without.

File: src/input_queue.rs [hunk 2]
```diff
@@ -11,9 +11,9 @@ pub(crate) struct InputQueue<T>
 where
     T: Config,
 {
-    /// The head of the queue. The newest `GameInput` is saved here      
+    /// The head of the queue. The newest `PlayerInput` is saved here
```
Comment (Vrixyz) [inline src/input_queue.rs:14]: Added renames from GameInput to PlayerInput (could be slit in another PR..)

File: tests/enum_input.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        let p1_inputs = inputs[1];
+
+        if p0_inputs == p1_inputs {
+            self.state += 2;
+        } else {
+            self.state -= 1;
+        }
+        self.frame += 1;
+    }
+}
```
Comment (Vrixyz) [inline tests/enum_input.rs:133]: I wrapped up tests in a very ugly fashion, just as a minimal check, I gotta clean this up, we probably should make a generic test over any input, and test over regular struct, and enums.

File: tests/enum_input.rs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    pub gs: StateStub,
+}
+use bytemuck::{CheckedBitPattern, NoUninit, Zeroable};
+
+#[repr(u8)]
+#[derive(Copy, Clone, PartialEq, CheckedBitPattern, NoUninit)]
+pub enum EnumInput {
+    Val1,
+    Val2,
+}
```
Comment (Vrixyz) [inline tests/enum_input.rs:24]: For now, it only supports fieldless enums, because NoUninit doesn't support fields, it might be technically possible to add this support upstream in bytemuck ?

File: tests/enum_input.rs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+pub struct StubConfig;
+
+impl Config for StubConfig {
+    type Input = EnumInput;
+    type State = StateStub;
+    type Address = SocketAddr;
+}
+
+impl GameStub {
```
Comment (Vrixyz) [inline tests/enum_input.rs:40]: I wrapped up tests in a very ugly fashion, just as a minimal check, I gotta clean this up, we probably should make a generic test over any input, and test over regular struct, and enums.

File: src/network/protocol.rs [hunk 6]
```diff
@@ -86,7 +86,7 @@ impl InputBytes {
         for p in 0..num_players {
             let start = p * size;
             let end = start + size;
-            let input = *bytemuck::try_from_bytes::<T::Input>(&self.bytes[start..end])
+            let input = *bytemuck::checked::try_from_bytes::<T::Input>(&self.bytes[start..end])
```
Comment (Vrixyz) [inline src/network/protocol.rs:89]: That's an important change.

File: tests/enum_input.rs (hunk 4 above)
Comment (Vrixyz) [reply tests/enum_input.rs:24]: I'm wondering if NoUninit is actually needed, and if yes, could MaybeUninit replace it ? 🤔 I'm gonna explore this idea.

Comment (gschup) [reply tests/enum_input.rs:24]: Please do! Sounds exciting! :)

File: src/network/protocol.rs (hunk 6 above)
Comment (gschup) [reply src/network/protocol.rs:89]: Can you elaborate what is actually checked here?

Comment (Vrixyz) [reply src/network/protocol.rs:89]: Sure, first, we need to understand the behaviour before this change:

With the restriction "AnyBitValid", enums could not work because an enum cannot have "any" bits ; mainly because its type flag, but probably other types are concerned (references, bool...)

Nothing is really checked if the type is "AnyBitPattern" though

the "checked" prefixed leads to: 
https://github.com/Lokathor/bytemuck/blob/6b1b7cecd32902b240a8e8593534600a09269d14/src/checked.rs#L209 ; where there is additional checks in case there are not valid bit patterns 😅.

All "AnyBitPatterns" implement "CheckedBitPattern" though, so no additional checks are performed for most types :).

File: src/frame_info.rs (hunk 1 above)
Comment (gschup) [reply src/frame_info.rs:35]: `bytemuck::Zeroable` is a key constraint used e.g. [here](https://github.com/gschup/ggrs/blob/0d70f4524b02a945999cdfed83a111e8c1140ed0/src/frame_info.rs#L46). 

The documentation on `bytemuck::NoUninit` recommends implementing `bytemuck::CheckedBitPattern` ([link](https://docs.rs/bytemuck/latest/bytemuck/trait.NoUninit.html)).

**Comment by gschup** [comment]
I like this. @Vrixyz what do you think: Is there anything else you want to add/change before we merge this?

**Comment by Vrixyz** [comment]
I'd like to make tests code quality better, I duplicated a lot of code and there may even be dead code.

A first step could be to:
- merge this once a minimal cleanup is done
- make a follow up issue to make more agressive improvements to the tests (generic tests over "CheckedBitPattern" ?)
- make an independent followup issue to attempt support for enums with fields.

File: tests/test_p2p_session.rs [hunk 7]
```diff
@@ -115,7 +115,7 @@ fn test_advance_frame_p2p_sessions() -> Result<(), GGRSError> {
     assert!(sess1.current_state() == SessionState::Synchronizing);
     assert!(sess2.current_state() == SessionState::Synchronizing);
 
-    for _ in 0..10 {
+    for _ in 0..50 {
```
Comment (Vrixyz) [inline tests/test_p2p_session.rs:118]: @gschup I would be happy to revert that, but on a 2015 macos, even `0..30` lead to sometimes failures. I figured it's a small price to pay to make the tests a little bit more stable. Eventually we would probably want a timeout, but it's too far from this Pull Request scope.

File: tests/test_synctest_session_enum.rs [hunk 8]
```diff
@@ -0,0 +1,25 @@
+mod stubs_enum;
+
+use ggrs::{GGRSError, SessionBuilder};
+
+#[test]
+fn test_enum_advance_frames_with_delayed_input() -> Result<(), GGRSError> {
```
Comment (Vrixyz) [inline tests/test_synctest_session_enum.rs:6]: as said in comment, I would have like to provide more tests, but it's creating a lot of duplicated code, so I figured we could merge minimal tests, and then discuss a better test strategy:
- I tried a generic `GameStub`, but all the data structures are quite tied, and the hardest to make generic is the `Config` implementation.
- macros could be another idea, but they can be controversial.

File: tests/enum_input.rs (hunk 4 above)
Comment (Vrixyz) [reply tests/enum_input.rs]: I gave it a small attempt, I think it's a good direction, but our problem is [bytes_of](https://docs.rs/bytemuck/latest/bytemuck/fn.bytes_of.html) used here https://github.com/gschup/ggrs/blob/0d70f4524b02a945999cdfed83a111e8c1140ed0/src/network/protocol.rs#L54, which has a type restriction on `NoUninit`. I *feel* like it should the contraint could be relaxed to support enum with fields, but it might need quite some work from bytemuck, such as https://github.com/Lokathor/bytemuck/issues/55.

I'm gonna create an issue on bytemuck and see where it goes from there.

File: Cargo.toml [hunk 9]
```diff
@@ -21,7 +21,7 @@ rand = "0.8"
 bitfield-rle = "0.2"
 parking_lot = "0.11"
 instant = "0.1"
-bytemuck = {version = "1.7", features = ["derive"]}
+bytemuck = {version = "1.9", features = ["derive"]}
```
Comment (Vrixyz) [inline Cargo.toml:24]: Pod trait was split in multiple traits in release 1.9.0 https://github.com/Lokathor/bytemuck/blob/main/changelog.md#190

**Comment by Vrixyz** [comment]
> I like this. @Vrixyz what do you think: Is there anything else you want to add/change before we merge this?

I'm done with this PR feature-wise, I'll address feedbacks only now :)

File: tests/enum_input.rs (hunk 4 above)
Comment (fu5ha) [reply tests/enum_input.rs]: Since I was reading thru this PR anyway, just a small note: `bytes_of` is used elsewhere so this doesn't solve the problem, but instead of creating a `zeroed` object just to call `bytes_of(&zeroed).len() * num_players` here, you could forego all of that and simply call `core::mem::size_of::<T::Input>() * num_players` to get the size

Comment (gschup) [reply tests/enum_input.rs]: thanks for the comment! good catch, that definitely seems like a smarter solution.

File: tests/test_p2p_session.rs (hunk 7 above)
Comment (gschup) [reply tests/test_p2p_session.rs:118]: I find this to be very odd. This never was an issue for me, but network packages are mysterious :D
